### PR TITLE
Fix CUDA watcher shutdown before shared-memory teardown

### LINKED ISSUE #521: [BUG] CUDA agent segfault at exit caused by watcher thread
Issue description:

When running GPU workloads (e.g. benchmark/gpu/workload/vec_add.cu) with the bpftime CUDA agent attached, the process sometimes segfaults on exit.
The crash happens in the CUDA watcher thread, which polls ctx->cuda_shared_mem->flag1 in bpf_attach_ctx::start_cuda_watcher_thread().

Root Cause: 

The agent process opens the global shared memory with bpftime_initialize_global_shm(shm_open_type::SHM_OPEN_ONLY).
In runtime/src/bpftime_shm_internal.cpp, the global destructor __destruct_shm() unconditionally calls bpftime_destroy_global_shm().

bpftime_shm::~bpftime_shm() unmaps the Boost shared memory segment and calls cudaHostUnregister(base_addr) for the whole segment, which includes the cuda::CommSharedMem holding flag1/flag2.

The CUDA watcher thread is started in bpf_attach_ctx::start_cuda_watcher_thread() and is detached; because bpf_attach_ctx is stored in a union (bpf_attach_ctx_holder) whose destructor is empty, ~bpf_attach_ctx() never runs at process exit, so cuda_watcher_should_stop is never set to true.

### PR DESCRIPTION
**Pull request by yunwei37**
Fixes #521.

## Problem

The CUDA watcher polls data inside the global shared-memory mapping. During teardown, `bpftime_destroy_global_shm()` could unmap that storage before the watcher stopped, causing the watcher to access unmapped memory during process exit.

## Change

- expose the existing idempotent watcher stop/join path inside the runtime
- invoke it immediately before the global shared-memory destructor unmaps the segment
- add one focused regression that starts a watcher, proves it handled a request, destroys shared memory while its attach context remains alive, and checks that no background access survives teardown

## Minimal scope

- 3 files, +32/-0
- production: +12/-0
- focused regression: +20
- keeps the existing per-context watcher, `CUDAContext` layout, constructor, registration, and lifecycle model unchanged
- no public header, API, configuration, dependency, fixture framework, or unrelated runtime fix
- replaces the previous process-global refcount/generation state machine, reducing the PR by 207 changed lines

## Validation

Exact HEAD: `41fd1d9d3acc2c78372fc1fedcc3a280157e9c93`

- synchronized with current `master`, including #626
- built `bpftime_runtime_tests` with CUDA attach enabled
- focused teardown regression: 1 assertion in 1 case passed
- focused teardown stress: 20/20 passed
- `git diff --check` passed
- changed-range clang-format produced no changes
- the unchanged neighboring `Test cuda compile` case fails standalone because the runtime VM name is empty; the watcher regression passes independently and the PR does not modify that test path

## Review and CI

Final-head subagent review and independent read-only review passed with no blockers. Copilot was explicitly requested; the final thread-aware audit found no applicable unresolved comment, and the two historical threads are resolved and outdated. GitHub Actions reached terminal green: 86 passed, 2 conditionally skipped, 0 failed, 0 cancelled, 0 pending. This PR is Ready for maintainer approval and must not be merged without it.

### CODE REVIEW

File: runtime/src/attach/bpf_attach_ctx_cuda.cpp [hunk 1]
```diff
@@ -273,46 +295,13 @@ void bpf_attach_ctx::start_cuda_watcher_thread()
...
-					cuda_watcher_thread.detach();
-			} catch (const std::system_error &detach_ex) {
-				SPDLOG_WARN(
-					"Failed to detach CUDA watcher thread after join failure: {}",
-					detach_ex.what());
-			}
-		}
-	}
+	(void)stop_registered_cuda_watcher(&cuda_watcher_thread);
 }
```
[the remark this answers is not in the document]
Comment (yunwei37) [reply runtime/src/attach/bpf_attach_ctx_cuda.cpp:321]: Ignoring the boolean here is intentional fail-stop behavior in the member destructor. On the only failure path, the helper preserves both registry and mapping and never detaches; the still-joinable std::thread prevents object destruction from silently proceeding into a UAF. Logging is not a safe recovery mechanism because this path can run during static teardown after spdlog state has been destroyed. The exact-head reviewers specifically verified no detach, preserved registration, and fail-closed mapping semantics.

File: runtime/src/bpftime_shm_internal.cpp [hunk 2]
```diff
@@ -48,6 +52,10 @@ extern "C" void bpftime_destroy_global_shm()
 {
 	using namespace bpftime;
 	if (global_shm_initialized) {
+#ifdef BPFTIME_ENABLE_CUDA_ATTACH
+		if (!stop_cuda_watcher_before_shm_unmap())
+			return;
+#endif
```
[the remark this answers is not in the document]
Comment (yunwei37) [reply runtime/src/bpftime_shm_internal.cpp:57]: The early return is deliberately fail-closed: if joining fails, unmapping would recreate the use-after-unmap this PR fixes. This function is also invoked from the static destructor, and the existing code immediately below documents that spdlog globals may already be destroyed at that point, so adding a teardown log here would itself be unsafe. The C API returns void and has no recoverable error channel; preserving the mapping is the only safe behavior on this exceptional path. Exact-head subagent and fresh-clone reviews both verified this tradeoff.

**Comment by yunwei37** [comment]
Current-head CI triage for `974d42d0ef83d02046796be8f4c5d7eae85125bc`: the representative GCC 13 failure is the pre-existing master syscall-server API rename at `runtime/syscall-server/syscall_context.cpp:75` (`construct_agent_config_from_env()` no longer exists). The one-line prerequisite is isolated in #626 and is not part of this PR. No speculative watcher change is warranted from that failure; after #626 merges I will sync/rerun and repeat exact-head review if the SHA changes.

AI-generated response; a maintainer will review and follow up later

**Comment by yunwei37** [comment]
Exact-head review refresh for `86e9e6e5980142cee6b00330e870e066dd3a653c`:

- Full subagent review: **PASS**. Clean worktree; merge-base `982edf825c92a1b2c303d153b7966cccd451fd12`; 4 files, +142/-58. CUDA targeted regression passed 100/100; non-CUDA agent/server build and diff checks passed.
- Independent Claude Opus maintainer review: **PASS**. It independently checked the create/start generation interleaving, shared watcher refcounts, stale-context destruction after reinitialization, atexit ordering, lock order, and minimality; the focused CUDA regression passed 20/20 on real hardware.
- Local exact-head validation: CUDA `bpftime_runtime_tests`, `bpftime-agent`, and `bpftime-syscall-server` built; focused regression passed; non-CUDA agent/server built; `git diff --check` passed.
- Both reviews found no blocking issue. Prior reviews on older commits are superseded.

The current GitHub build failures are the known master-side syscall-server API rename fixed by prerequisite #626; this PR intentionally does not duplicate that unrelated one-line fix. Keeping this PR Draft until #626 merges, current-head CI is rerun to a terminal green state, and the refreshed Copilot review is closed.

**Comment by yunwei37** [comment]
@copilot review

**Comment by yunwei37** [comment]
@copilot review

**Comment by yunwei37** [comment]
@copilot review

Please review exact HEAD 8b34c1e583fa32c67f1a165774912c4d791556c6, including the final unique-pointer ownership minimality change.

**Comment by yunwei37** [comment]
Exact-head review evidence for 8b34c1e583fa32c67f1a165774912c4d791556c6

- Dedicated subagent review: PASS, no code blocker.
- Independent external read-only review: PASS, no practical blocker.
- Local HEAD, upstream, and PR head all matched before review.
- Minimality: PASS. The final ownership refinement is one file, +2/-2, replacing shared ownership with a unique-pointer move-capture. Full scope remains 4 files, +143/-59, limited to watcher lifecycle coordination and one focused regression.
- Lifecycle checks covered global watcher ownership, user and generation accounting, stale contexts, join-before-unmap, destroy and reinitialize, atexit, concurrent contexts, stop failure behavior, and move-only thread capture.
- Validation: bpftime_runtime_tests, bpftime-agent, and bpftime-syscall-server built; the focused CUDA lifecycle test passed 20 consecutive runs and a visible 2-assertion run; diff and format checks passed.
- The external reviewer noted only theoretical uint64 generation wrap after roughly 2^64 teardown cycles; this is not practical or blocking.

This is code-review clearance only. Current CI is still running and the branch must be synchronized after #626 before final current-base CI and Copilot gates.

**Comment by yunwei37** [comment]
@copilot review

Please review exact HEAD 386cbe95124958e3b87f67acdc2d2de4cb8bc5c8 after the final minimality and formatting reduction.

**Comment by yunwei37** [comment]
@copilot review

**Comment by yunwei37** [comment]
@copilot review

**Comment by yunwei37** [comment]
## Final-head review evidence — PASS

Exact head: `41fd1d9d3acc2c78372fc1fedcc3a280157e9c93`

- Dedicated subagent review: PASS, no blockers.
- Independent Claude read-only review: PASS, no blockers.
- Both reviewers blocked the previous head because its process-global refcount/generation state machine and `CUDAContext` layout/constructor changes exceeded issue #521.
- Commit `41fd1d9` removes that state machine and restores the existing per-context watcher model and public header unchanged.
- Final diff: 3 files, +32/-0; production +12, focused test +20. This is 207 changed lines smaller than the previous head.
- The root fix reuses the existing idempotent stop/join path immediately before `bpftime_shm` destruction and unmapping; `__destruct_shm()` reaches the same path.
- The focused test proves a watcher is active, destroys shared memory while its context remains alive, and holds the context for five watcher periods. It passed once visibly and 20/20 under repetition.
- Both reviewers found no test/fixture bloat and no unauthorized API/ABI, default, documentation, or architecture change.
- `git diff --check` and changed-range clang-format passed.

Copilot review and GitHub Actions are still pending; this is not a merge-complete claim.

**Comment by yunwei37** [comment]
## Final OSS gate — READY FOR MAINTAINER APPROVAL

Exact head: `41fd1d9d3acc2c78372fc1fedcc3a280157e9c93`

- final-head maintainer review: PASS
- independent external read-only review: PASS
- Copilot explicitly requested; final thread-aware audit found no applicable unresolved comments (2 historical threads are resolved and outdated)
- GitHub Actions terminal: 86 passed, 2 conditionally skipped, 0 failed, 0 cancelled, 0 pending
- focused local regression: 1/1 passed; stress: 20/20 passed
- final diff: 3 files, `+32/-0`; production `+12`, test `+20`; 207 changed lines smaller than the previous head
- PR is non-draft and mergeable; branch protection still requires maintainer review

No merge was performed.
